### PR TITLE
Refactoring of CoverCrypt API

### PR DESCRIPTION
**Pull request by Adamk93**

### CODE REVIEW

File: src/core/api.rs [hunk 1]
```diff
@@ -40,13 +40,299 @@ impl PartialEq for Covercrypt {
...
+    /// MPK.
+    ///
+    /// Sets the MPK coordinates to the one defined by the policy:
+    /// - removes coordinates from the MSK that don't belong to the new policy
+    /// along with their associated keys;
+    /// - adds the policy coordinates that don't belong yet to the MSK,
+    /// generating new keys.
+    ///
+    /// The new MPK holds the latest public keys of each coordinates of the new policy.
+    fn update_master_keys(
```
Comment (tbrezot) [inline src/core/api.rs:237]: key management should not be part of the KEM interface.

File: src/core/api.rs [hunk 2]
```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
...
         usk: &UserSecretKey,
-        encapsulation: &Encapsulation,
-    ) -> Result<Option<SymmetricKey<SEED_LENGTH>>, Error> {
-        decaps(usk, encapsulation)
+        encapsulation: Encapsulation,
+    ) -> Option<Secret<SEED_LENGTH>> {
+        match decaps(usk, &encapsulation) {
+            Err(_) => None,
+            Ok(sk) => Some(sk?),
+        }
```
Comment (tbrezot) [inline src/core/api.rs:513]: No you should keep a `Result<Option<...>, Error>`: the error semantic is not the same as the option one. You get an error in case something went wrong; you get `None` if the USK rights did not match the encapsulation's access policy.

File: src/core/api.rs [hunk 3]
```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
...
-        let nonce = Nonce::new(&mut *self.rng.lock().expect("could not lock mutex"));
-        let mut ciphertext = aes256gcm.encrypt(&nonce, plaintext, ad)?;
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
-        res.append(&mut ciphertext);
-        Ok(res)
-    }
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error>;
```
Comment (tbrezot) [inline src/core/api.rs:540]: I wonder if we should already serialize the encapsulation: the current solution seems to be somewhere in-between returning a struct containing each parts (encapsulation, mac, none and ciphertext) and serializing and concatenating all parts.

I don't have any strong opinion on it but this solution is surprising.

File: src/core/api.rs [hunk 4]
```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
...
-                encrypted_metadata,
-            },
-        ))
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error> {
+        let (seed, enc) = CovercryptKEM::encaps(self, mpk, policy, ap)?;
+        let mut sym_key = Secret::<KEY_LENGTH>::default();
+        kdf256!(&mut sym_key, &seed);
```
Comment (tbrezot) [inline src/core/api.rs:586]: The hash can be avoided if we truncate. Anyway, there is not enough entropy to create a key of greater length:
```suggestion
        if SEED_LENGTH < Self::KEY_LENGTH {
            return Err(...);
        }
        let key = SymmetricKey::try_from_slice(&seed[..Self::KEY_LENGTH])?;
```

File: src/core/api.rs [hunk 5]
```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
...
+        ad: Option<&[u8]>,
+        ciphertext: &[u8],
+        enc: &Encapsulation,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let seed = CovercryptKEM::decaps(self, usk, enc.clone());
+        seed.map(|seed| {
+            let mut sym_key = Secret::<KEY_LENGTH>::default();
+            kdf256!(&mut sym_key, &seed);
+            let plaintext = E::decrypt(&sym_key, ad, ciphertext)?;
+            Ok(plaintext)
```
Comment (tbrezot) [inline src/core/api.rs:612]: Maybe the error conversion can be inferred:
```suggestion
            E::decrypt(&sym_key, ad, ciphertext).into()
```

**Review by tbrezot — COMMENTED** [review]
Avoid moving lines around: it makes diff really hard to follow. Otherwise great work.

File: examples/runme.rs [hunk 6]
```diff
@@ -42,48 +44,73 @@ fn main() {
 
     // Setup Covercrypt and generate master keys
     let cover_crypt = Covercrypt::default();
-    let (mut msk, _) = cover_crypt.setup().unwrap();
-    let mpk = cover_crypt.update_master_keys(&policy, &mut msk).unwrap();
+    let (mut msk, _) = CovercryptKEM::setup(&cover_crypt).unwrap();
+    let mpk = CovercryptKEM::update_master_keys(&cover_crypt, &policy, &mut msk).unwrap();
```
[the remark this answers is not in the document]
Comment (Adamk93) [reply examples/runme.rs:48]: should they be a part of the PKE trait?

File: examples/decrypt.rs [hunk 7]
```diff
@@ -1,3 +1,5 @@
+use cosmian_crypto_core::Aes256Gcm;
```
[the remark this answers is not in the document]
Comment (Adamk93) [reply examples/decrypt.rs:1]: I agree, but I did not change the semantics of the original example, and it was related to the EncryptionHeader. What I can do, is to expose the encryption header already instantiated with Aes256Gcm

File: src/core/api.rs (hunk 1 above)
Comment (Adamk93) [reply src/core/api.rs:237]: Similarly: should it be in PKE?

File: examples/runme.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbrezot) [reply examples/runme.rs:48]: No, key-management is a Covercrypt-specific thing, it should be implemented for the Covercrypt type onyl.

**Comment by Adamk93** [comment]
Missing only implementation of theEncryptionHandler methods. Now they as trait's functions

**Comment by Adamk93** [comment]
@tbrezot I think I've addressed all your comments. Maybe miscommunication, but idk if I have to do a new pull request or not.

File: src/core/api.rs [hunk 8]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
+        let nonce = &self::Nonce::<{ Self::NONCE_LENGTH }>::new(&mut *rng);
+        let aes = Self::new(key);
+        let mut ciphertext = aes
+            .encrypt(nonce, ptx, ad)
+            .map_err(Error::CryptoCoreError)?;
+        let mut res = Vec::with_capacity(ptx.len() + Self::MAC_LENGTH + Self::NONCE_LENGTH);
+        res.extend(&Nonce::to_bytes(nonce));
```
Comment (tbrezot) [inline src/core/api.rs:232]: This should work and avoids a copy:
```suggestion
        res.extend(&nonce);
```

File: src/core/api.rs [hunk 9]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
-        let mut ciphertext = aes256gcm.encrypt(&nonce, plaintext, ad)?;
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
+        let nonce = &self::Nonce::<{ Self::NONCE_LENGTH }>::new(&mut *rng);
+        let aes = Self::new(key);
+        let mut ciphertext = aes
+            .encrypt(nonce, ptx, ad)
+            .map_err(Error::CryptoCoreError)?;
+        let mut res = Vec::with_capacity(ptx.len() + Self::MAC_LENGTH + Self::NONCE_LENGTH);
```
Comment (tbrezot) [inline src/core/api.rs:231]: This is more explicit to use the ctx length:
```suggestion
        let mut res = Vec::with_capacity(ctx.len() + Self::NONCE_LENGTH);
```
Alternatively, you could use:
```rust
return [&nonce, &ctx].concat();
```

File: src/core/api.rs [hunk 10]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
-    ///
-    /// - `symmetric_key`   : AES key
-    /// - `ciphertext`      : encrypted data
-    /// - `ad`              : associated data
-    pub fn dem_decrypt(
-        &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        ciphertext: &[u8],
+    fn decrypt(
+        key: &Secret<{ Self::KEY_LENGTH }>,
```
Comment (tbrezot) [inline src/core/api.rs:238]:
```suggestion
        key: &SymmetricKey<{ Self::KEY_LENGTH }>,
```

File: src/core/api.rs [hunk 11]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
-                &Nonce::try_from_slice(&ciphertext[..Aes256Gcm::NONCE_LENGTH])?,
-                &ciphertext[Aes256Gcm::NONCE_LENGTH..],
-                ad,
-            )
+        if ctx.len() < Self::NONCE_LENGTH {
+            return Err(Error::CryptoCoreError(
+                cosmian_crypto_core::CryptoCoreError::DecryptionError,
+            ));
+        }
+        let sym_key = SymmetricKey::try_from_slice(key)?;
```
Comment (tbrezot) [inline src/core/api.rs:247]: This type conversion should have given you a hint.
```suggestion
```

File: src/core/api.rs [hunk 12]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
+        let nonce =
+            &Nonce::<{ Self::NONCE_LENGTH }>::try_from_slice(&ctx[..{ Self::NONCE_LENGTH }])?;
```
Comment (tbrezot) [inline src/core/api.rs:249]: These curly braces should not be needed, the nonce length can be inferred and avoid taking only refs on your variables (it already tricked you into unnecessarily cloning).
```suggestion
            Nonce::try_from_slice(&ctx[..Self::NONCE_LENGTH])?;
```

File: src/core/api.rs [hunk 13]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
+        let aes = Self::new(&sym_key);
+        aes.decrypt(nonce, &ctx[{ Self::NONCE_LENGTH }..], ad)
```
Comment (tbrezot) [inline src/core/api.rs:251]:
```suggestion
        aes.decrypt(&nonce, &ctx[Self::NONCE_LENGTH..], ad)
```

File: src/core/api.rs [hunk 14]
```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
...
-pub struct EncryptedHeader {
+pub struct EncryptedHeader<
+    E: AE<KEY_LENGTH, NONCE_LENGTH, MAC_LENGTH>,
+    const KEY_LENGTH: usize,
+    const NONCE_LENGTH: usize,
+    const MAC_LENGTH: usize,
+> {
     pub encapsulation: Encapsulation,
     pub encrypted_metadata: Option<Vec<u8>>,
+    pub phantom: PhantomData<(E, usize, usize, usize)>,
```
Comment (tbrezot) [inline src/core/api.rs:268]: You only need to use generic types:
```suggestion
    pub phantom: PhantomData<E>,
```

File: src/core/api.rs [hunk 15]
```diff
@@ -279,22 +289,30 @@ impl EncryptedHeader {
         cover_crypt: &Covercrypt,
         policy: &Policy,
         public_key: &MasterPublicKey,
-        encryption_policy: &AccessPolicy,
+        encryption_policy: &str,
```
Comment (tbrezot) [inline src/core/api.rs:292]: I thought you were in favor of reverting the move from `AccessPolicy` to `String`?

File: src/core/api.rs [hunk 16]
```diff
@@ -312,21 +330,23 @@ impl EncryptedHeader {
...
-                })
-                .transpose()?;
-            Ok(Some(CleartextHeader {
-                symmetric_key,
-                metadata,
-            }))
-        } else {
-            Ok(None)
-        }
+        CovercryptKEM::decaps(cover_crypt, usk, self.encapsulation.clone()).map(|seed| {
```
Comment (tbrezot) [inline src/core/api.rs:333]: No clone needed: `CovercryptKem` does not need taking ownership of the encapsulation.
```suggestion
        CovercryptKEM::decaps(cover_crypt, usk, &self.encapsulation).map(|seed| {
```

File: src/core/api.rs [hunk 17]
```diff
@@ -312,21 +330,23 @@ impl EncryptedHeader {
...
+            if let Some(mut seed) = seed {
```
Comment (tbrezot) [inline src/core/api.rs:334]: I would rather like you to generate a new seed instead.
```suggestion
            if let Some(seed) = seed {
```

File: src/core/api.rs [hunk 18]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
+    fn decaps(
+        &self,
+        usk: &UserSecretKey,
+        enc: Encapsulation,
```
Comment (tbrezot) [inline src/core/api.rs:385]:
```suggestion
        enc: &Encapsulation,
```

File: src/core/api.rs [hunk 19]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error>; // todo : symmetric_key or seed?
+
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
```
Comment (tbrezot) [inline src/core/api.rs:380]:
```suggestion
    /// Returns the encapsulated seed if the user key holds
```

File: src/core/api.rs [hunk 20]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encaps(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error>; // todo : symmetric_key or seed?
```
Comment (tbrezot) [inline src/core/api.rs:375]: Is this still an open question?

File: src/core/api.rs [hunk 21]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    ) -> Result<Option<Secret<SEED_LENGTH>>, Error>;
+}
+
+impl CovercryptKEM for Covercrypt {
+    /// Generates an encapsulation for the given access
+    /// policy.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
```
Comment (tbrezot) [inline src/core/api.rs:395]: No need to repeat the doc: it will be derived from the trait.
```suggestion
```

File: src/core/api.rs [hunk 22]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error> {
+        let ap = &AccessPolicy::parse(ap)?;
+        encaps(
+            &mut *self.rng.lock().expect("Mutex lock failed!"),
+            mpk,
+            &policy.generate_point_coordinates(ap.clone())?,
+        )
```
Comment (tbrezot) [inline src/core/api.rs:407]: Another example of using a variable to store a reference tricking you into a useless clone.
```suggestion
        let ap = AccessPolicy::parse(ap)?;
        encaps(
            &mut *self.rng.lock().expect("Mutex lock failed!"),
            mpk,
            &policy.generate_point_coordinates(ap)?,
        )
```

File: src/core/api.rs [hunk 23]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    }
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
```
Comment (tbrezot) [inline src/core/api.rs:413]:
```suggestion

```

File: src/core/api.rs [hunk 24]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    fn decaps(
+        &self,
+        usk: &UserSecretKey,
+        encapsulation: Encapsulation,
+    ) -> Result<Option<Secret<SEED_LENGTH>>, Error> {
+        decaps(usk, &encapsulation)
```
Comment (tbrezot) [inline src/core/api.rs:419]:
```suggestion
        encapsulation: &Encapsulation,
    ) -> Result<Option<Secret<SEED_LENGTH>>, Error> {
        decaps(usk, encapsulation)
```

File: src/core/api.rs [hunk 25]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    const KEY_LENGTH: usize,
+    const NONCE_LENGTH: usize,
+    const MAC_LENGTH: usize,
+>
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
+    ///
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
```
Comment (tbrezot) [inline src/core/api.rs:434]:
```suggestion
    /// Creates a Covercrypt encapsulation of a 256-bit seed, and use it with
    /// the given authentication data to encrypt the plaintext.
```

File: src/core/api.rs [hunk 26]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+    /// authentication data. Returns ‘None‘ otherwise.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn decrypt(
+        &self,
+        usk: &UserSecretKey,
+        ad: Option<&[u8]>,
+        cyphertext: &[u8],
```
Comment (tbrezot) [inline src/core/api.rs:461]:
```suggestion
        ciphertext: &[u8],
```
or `ctx`

File: src/core/api.rs [hunk 27]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
+    ///
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
```
Comment (tbrezot) [inline src/core/api.rs:481]:
```suggestion
```

File: src/core/api.rs [hunk 28]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error> {
+        let (seed, enc) = CovercryptKEM::encaps(self, mpk, policy, ap)?;
+        if seed.length() < KEY_LENGTH {
+            return Err(Error::KeyError(
+                String::from_str("SEED_LENGTH >= KEY_LENGTH").unwrap(),
+            ));
+        }
```
Comment (tbrezot) [inline src/core/api.rs:495]: - error message is wrong or at least misleading;
- just rehash the seed instead

File: src/core/api.rs [hunk 29]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+        let sym_key = SymmetricKey::<KEY_LENGTH>::try_from_slice(&seed[..KEY_LENGTH])?;
```
Comment (tbrezot) [inline src/core/api.rs:496]: We may want to implement a key derivation from a secret:
```rust
impl<const KEY_LENGTH: usize> SymmetricKey<KEY_LENGTH> {
    fn derive<const SEED_LENGTH: usize>(&mut impl CryptoRngCore, seed: &Seed<SEED_LENGTH>, info: &[u8]) -> Result<Self, Error> {
        if SEED_LENGTH < KEY_LENGTH {
            Err(...) // insufficient entropy
        } else {
            let mut key = Self::default();
            kdf!(&mut key, seed, info);
            Ok(key)
        }
    }
}```

File: src/core/api.rs [hunk 30]
```diff
@@ -336,6 +356,188 @@ impl EncryptedHeader {
...
+
+    /// Attempts decrypting the given ciphertext using the Covercrypt KEM and the DEM.
+    ///
+    /// Attempts opening the Covercrypt encapsulation. If it succeeds, decrypts
+    /// the ciphertext using the DEM with the encapsulated key and the given
+    /// authentication data. Returns ‘None‘ otherwise.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
```
Comment (tbrezot) [inline src/core/api.rs:510]:
```suggestion
```

**Review by tbrezot — COMMENTED** [review]
Almost there, good job :+1:

File: src/core/api.rs (hunk 8 above)
Comment (Adamk93) [reply src/core/api.rs:232]: extend needs an iterable type. Nonce does not have the trait implemented.

File: src/core/api.rs (hunk 29 above)
Comment (Adamk93) [reply src/core/api.rs:496]: This has to be implemented in crypto_core, right? The type is not defined in Covercrypt. @tbrezot

Comment (tbrezot) [reply src/core/api.rs]: yes

File: src/core/api.rs (hunk 8 above)
Comment (tbrezot) [reply src/core/api.rs]: OK, my bad, Idk why but `Deref` is not implemented for this type. Then you should use `as_bytes()` instead of `to_bytes()` to avoid a copy.
